### DISCUSSION TITLE
i cant understand addressSeed

### DISCUSSION THREAD

**Discussion by YavorJJ** [Cyfrin/foundry-full-course-cu#2030] [Q&A] [answered]
Hello, I have been really stuck tryna understand addressSeed and line of code here: address sender = userswithCollateralDeposited[addressSeed % userswithCollateralDeposited.length];  this is in handler contract in the stablecoin defi project. Lesson 23. 

` ` `solidity  //SPDX-License-Idnetifier: MIT
    // Handlar is going to narrow down the way we call the function

    pragma solidity ^0.8.19;

    import {Test} from "forge-std/Test.sol";
    import {DSCEngine} from "../../src/DSCEngine.sol";
    import {DecentralizedStableCoin} from "../../src/DecentralizedStableCoin.sol";
    import {ERC20Mock} from "@openzeppelin/contracts/mocks/token/ERC20Mock.sol";

    contract Handler is Test {
        DSCEngine dsce;
        DecentralizedStableCoin dsc;

        ERC20Mock weth;
        ERC20Mock wbtc;

        uint256 public timesMintIsCalled;
        address[] public userswithCollateralDeposited;

        uint256 MAX_DEPOSIT_SIZE = type(uint96).max;

        constructor (DSCEngine _dscEngine, DecentralizedStableCoin _dsc) {
            dsce = _dscEngine;
            dsc = _dsc;

            address[] memory collateralTokens = dsce.getCollateralTokens();
            weth = ERC20Mock(collateralTokens[0]);
            wbtc = ERC20Mock(collateralTokens[1]);
        }    

        function mintDsc (uint256 amount, uint256 addressSeed) public {
            if(userswithCollateralDeposited.length == 0) {
                return;
            }
            address sender = userswithCollateralDeposited[addressSeed % userswithCollateralDeposited.length];
            (uint256 totalDscMinted, uint256 collateralValueInUsd) = dsce.getAccountInformation(sender);    

            int256 maxDscToMint = (int256(collateralValueInUsd) / 2) - int256(totalDscMinted); 
            if (maxDscToMint < 0) {
                return;
            }
            amount = bound(amount, 0, uint256(maxDscToMint));
            if (amount == 0) {
                return;
            }
            vm.startPrank(sender);
            dsce.mintDsc(amount);
            vm.stopPrank;
            timesMintIsCalled++;
        }

        // redeem collateral <-

        function depositCollateral(uint256 collateralSeed, uint256 amountCollateral) public {
            ERC20Mock collateral = _getCollateralFromSeed(collateralSeed);
            amountCollateral = bound(amountCollateral, 1, MAX_DEPOSIT_SIZE);
            
            vm.startPrank(msg.sender);
            collateral.mint(msg.sender, amountCollateral);
            collateral.approve(address(dsce), amountCollateral);
            dsce.depositCollateral(address(collateral), amountCollateral);
            vm.stopPrank();
            // this can double push if the same address is pushed twice.
            userswithCollateralDeposited.push(msg.sender);

        }   

        function redeemCollateral(uint256 collateralSeed, uint256 amountCollateral) public {
            ERC20Mock collateral = _getCollateralFromSeed(collateralSeed);
            uint256 maxCollateralToRedeem = dsce.getCollateralBalanceOfUser(address(collateral), msg.sender);
            amountCollateral = bound(amountCollateral, 0, maxCollateralToRedeem);
            if (amountCollateral == 0) {
                return;
            }
            dsce.redeemCollateral(address(collateral), amountCollateral);
        }

        // Helper Functions
        function _getCollateralFromSeed(uint256 collateralSeed) private view returns (ERC20Mock) {
              if (collateralSeed % 2 == 0){  
                  return weth;
              }
              return wbtc;
        }

    }
` ` ` solidity

**Comment by EngrPips** [comment, marked as answer]
Hello @YavorJJ, I want us to remember that we are doing fuzz testing, meaning that the fuzzer will call random functions with random values to try and break our system. Still, if we have a complex system like `DSCEngine,` then the randomness of the fuzzer will make no sense if we don't guide and control it. The line of code that confused you, `address sender = userswithCollateralDeposited[addressSeed % userswithCollateralDeposited.length]`, is inside the `mintDsc` function of our`Handler` contract. If you have been following along appropriately, you will know that anyone who will mint `Dsc` token must have initially deposited some collateral to the `DSCEngine`. Let's look at the `mintDsc` function inside our `Handler` contract to understand what it does.

```js
function mintDsc (uint256 amount, uint256 addressSeed) public {
        if(userswithCollateralDeposited.length == 0) {
            return;
        }
        address sender = userswithCollateralDeposited[addressSeed % userswithCollateralDeposited.length];
        (uint256 totalDscMinted, uint256 collateralValueInUsd) = dsce.getAccountInformation(sender);    

        int256 maxDscToMint = (int256(collateralValueInUsd) / 2) - int256(totalDscMinted); 
        if (maxDscToMint < 0) {
            return;
        }
        amount = bound(amount, 0, uint256(maxDscToMint));
        if (amount == 0) {
            return;
        }
        vm.startPrank(sender);
        dsce.mintDsc(amount);
        vm.stopPrank;
        timesMintIsCalled++;
    }
```
The function above takes `addressSeed` as the second argument instead of `address` because we don't want the fuzzer to give us a random address that most likely hasn't deposited any collateral to our `DSCEngine.` When the fuzzer gives any random value as `addressSeed,` then our `mintDsc` uses this value to pick from an array of addresses that we have been updating every time any address deposits some collateral to our `DSCEngine` by using the below line of code. So, any address we choose from our array of addresses will be eligible to mint `Dsc` within the limit of what they are eligible to mint according to `DSCEngine` predefined rules.
```js
address sender = userswithCollateralDeposited[addressSeed % userswithCollateralDeposited.length]
```

Please ask questions if anything is confusing.

**Reply by YavorJJ** [reply]
Appreciate it, thank you that you are there to help me, wish you the best. I understand it now.

**Reply by EngrPips** [reply]
Awesome. I am glad you figured it out. Keep crushing it.
